### PR TITLE
feat(avro-schema): logical type support

### PR DESCRIPTION
**Pull request by joscha**
This implements [logical type](https://avro.apache.org/docs/++version++/specification/#logical-types) support for:
- [`timestamp`](https://avro.apache.org/docs/++version++/specification/#timestamps)
- [`date`](https://avro.apache.org/docs/++version++/specification/#date)

for the OpenAPI [built-in `string` formats](https://swagger.io/docs/specification/data-models/data-types/):
- `date-time`
- `date`

The change is backwards compatible by allowing to opt-in to the behaviour via a new property `useLogicalTypes`.
Avro's support for `timestamp` fields includes the definition which timescale they are represented in (milliseconds, microseconds or nanoseconds). This can be controlled via a new property `logicalTypeTimeQuantifier`.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] ~If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.~ There is no committee for this one, as it's a schema transformer, thus: @sgadouar or @wing328 please?

### CODE REVIEW

File: bin/configs/avro-schema.yaml [hunk 1]
```diff
@@ -2,3 +2,6 @@ generatorName: avro-schema
 outputDir: samples/openapi3/schema/petstore/avro-schema
 inputSpec: modules/openapi-generator/src/test/resources/3_0/avro-schema/petstore.yaml
 templateDir: modules/openapi-generator/src/main/resources/avro-schema
+additionalProperties:
+  useLogicalTypes: true
+  logicalTypeTimeQuantifier: nanos
```
Comment (joscha) [inline bin/configs/avro-schema.yaml:7]:
```suggestion
  logicalTypeTimeQuantifier: nanos

```

File: modules/openapi-generator/src/main/resources/avro-schema/dataType.mustache [hunk 2]
```diff
@@ -0,0 +1 @@
+{{#useLogicalTypes}}{{#isDate}}{ "type": "int", "logicalType": "date" }{{/isDate}}{{#isDateTime}}{ "type": "long", "logicalType": "timestamp-{{{logicalTypeTimeQuantifier}}}" }{{/isDateTime}}{{^isDate}}{{^isDateTime}}"{{{dataType}}}"{{/isDateTime}}{{/isDate}}{{/useLogicalTypes}}{{^useLogicalTypes}}"{{{dataType}}}"{{/useLogicalTypes}}
```
Comment (joscha) [inline modules/openapi-generator/src/main/resources/avro-schema/dataType.mustache:1]: would be great to format this better; if there is a way to use whitespace in the template and then make it collapse in the result, please let me know.

Comment (wing328) [reply modules/openapi-generator/src/main/resources/avro-schema/dataType.mustache:1]: are you looking for something like following?

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Javascript/libraries/javascript/api.mustache#L3
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/Javascript/libraries/javascript/api.mustache#L118

Comment (joscha) [reply modules/openapi-generator/src/main/resources/avro-schema/dataType.mustache:1]: no, that's changing the template operator, right? I basically wish I could change the string in this template to:
```suggestion
{{#useLogicalTypes}}
  {{#isDate}}
    {
      "type": "int",
      "logicalType": "date"
    }
  {{/isDate}}
  {{#isDateTime}}
    {
     "type": "long",
     "logicalType": "timestamp-{{{logicalTypeTimeQuantifier}}}"
    }
  {{/isDateTime}}
  {{^isDate}}
    {{^isDateTime}}
      "{{{dataType}}}"
    {{/isDateTime}}
  {{/isDate}}
{{/useLogicalTypes}}
{{^useLogicalTypes}}
  "{{{dataType}}}"
{{/useLogicalTypes}}
```
and still have well-spaced output in the resulting schema. Something like a `{{#collapseWhitespace}}...{{/collapseWhitespace}}` operator that takes anything in its body and reduces all `\w+` to ` `

Comment (wing328) [reply modules/openapi-generator/src/main/resources/avro-schema/dataType.mustache:1]: nope. i don't think mustache comes with anything to do that

looks like avsc file are just JSON file so users can use a linter to auto format the auto-generated output as they do with other generators

**Pull request merged**